cg/services/order_validation_service/rules/case_sample/utils.py

```diff
@@ -22,3 +51,252 @@ def get_invalid_panels(panels: list[str], store: Store) -> list[str]:
         panel for panel in panels if not store.does_gene_panel_exist(panel)
     ]
     return invalid_panels
+
+
+def is_volume_invalid(sample: Sample) -> bool:
+    in_container: bool = is_in_container(sample.container)
+    allowed_volume: bool = is_volume_within_allowed_interval(sample.volume)
+    return in_container and not allowed_volume
+
+
+def is_in_container(container: ContainerEnum) -> bool:
+    return container != ContainerEnum.no_container
+
+
+def is_volume_within_allowed_interval(volume: int) -> bool:
+    return MINIMUM_VOLUME <= volume <= MAXIMUM_VOLUME
+
+
+def get_well_sample_map(
+    order: OrderWithCases, **kwargs
+) -> dict[tuple[str, str], list[tuple[int, int]]]:
+    """
+    Constructs a dict with keys being a (container_name, well_position) pair. For each such pair, the value will be
+    a list of (case index, sample index) pairs corresponding to all samples with matching container_name and
+    well_position, provided the sample is on a plate.
+    """
+    well_position_to_sample_map = {}
+    for case_index, case in order.enumerated_new_cases:
+        for sample_index, sample in case.enumerated_new_samples:
+            if is_sample_on_plate(sample):
+                key: tuple[str, str] = (sample.container_name, sample.well_position)
+                value: tuple[int, int] = (case_index, sample_index)
+                if not well_position_to_sample_map.get(key):
+                    well_position_to_sample_map[key] = []
+                well_position_to_sample_map[key].append(value)
+    return well_position_to_sample_map
+
+
+def get_occupied_well_errors(colliding_samples: list[tuple[int, int]]) -> list[OccupiedWellError]:
+    errors: list[OccupiedWellError] = []
+    for sample_index, case_index in colliding_samples:
+        error = OccupiedWellError(case_index=case_index, sample_index=sample_index)
+        errors.append(error)
+    return errors
+
+
+def is_sample_on_plate(sample: SampleWithRelatives) -> bool:
+    return sample.container == ContainerEnum.plate
+
+
+def get_indices_for_repeated_case_names(order: OrderWithCases) -> list[int]:
+    counter = Counter([case.name for case in order.cases])
+    indices: list[int] = []
+
+    for index, case in order.enumerated_new_cases:
+        if counter.get(case.name) > 1:
+            indices.append(index)
+
+    return indices
+
+
+def get_repeated_case_name_errors(order: OrderWithCases) -> list[RepeatedCaseNameError]:
+    case_indices: list[int] = get_indices_for_repeated_case_names(order)
+    return [RepeatedCaseNameError(case_index=case_index) for case_index in case_indices]
+
+
+def get_indices_for_repeated_sample_names(case: CaseContainingRelatives) -> list[int]:
```

Comment:
Could this not make do with a `Case` typehint?

Reply:
https://github.com/Clinical-Genomics/cg/pull/3625/commits/d4639b908ce95a8fe2881019e3934bd844f94afb